java/ql/src/experimental/Security/CWE/CWE-502/RmiUnsafeDeserialization.ql

```diff
@@ -0,0 +1,65 @@
+/**
+ * @name Unsafe deserialization with RMI.
+ * @description If a registered remote object has a method that accepts a complex object,
+ *              an attacker can take advantage of the unsafe deserialization mechanism
+ *              which is used to pass parameters in RMI.
+ *              In the worst case, it results in remote code execution.
+ * @kind problem
+ * @problem.severity error
+ * @precision high
+ * @id java/unsafe-deserialization-rmi
+ * @tags security
+ *       external/cwe/cwe-502
+ */
+
+import java
+import semmle.code.java.frameworks.Rmi
+
+private class ObjectInputStream extends RefType {
+  ObjectInputStream() { hasQualifiedName("java.io", "ObjectInputStream") }
+}
+
+/**
+ * A method that binds a name to a remote object.
+ */
+private class BindMethod extends Method {
+  BindMethod() {
+    (
+      getDeclaringType().hasQualifiedName("java.rmi", "Naming") or
+      getDeclaringType().hasQualifiedName("java.rmi.registry", "Registry")
+    ) and
+    hasName(["bind", "rebind"])
+  }
+}
+
+/**
+ * Looks for a vulnerable method in a `Remote` object.
+ */
+private Method getVulnerableMethod(Type type) {
+  type.(RefType).getASupertype*() instanceof TypeRemote and
```

Comment:
If I understand the documentation of [`Remote`](https://docs.oracle.com/en/java/javase/16/docs/api/java.rmi/java/rmi/Remote.html) correctly, only methods declared in an interface which directly extends `Remote` are remote methods.
Therefore the following would probably not be vulnerable (though I am not sure):
```java
interface RemoteInterface extends Remote {
  void doSomething(int);
}

class RemoteClass implements RemoteInterface {
  @Override
  public void doSomething(int) {}

  // Not vulnerable
  public void doSomething(Object o) {}
}
```

Reply:
Good catch! I'll update the query. Thanks!